### PR TITLE
EPMRPP-84613 Playwright tags are parsed to the RP's attributes and attached to the tests

### PR DESCRIPTION
**Pull request by abdelrhman-arnos**
The tags from Playwright should be converted into ReportPortal attributes and associated with the respective tests. If the test name carries tags as per the Playwright annotation conventions, these tags need to be included as test attributes in ReportPortal. Conversely, the test name itself should appear in ReportPortal without these tags. 

For example, a test titled `Test login page @tag` should appear in ReportPortal as 'Test login page' with an attribute that reads `{ value: 'tag' }`. Any tests that do not carry tags should be reported in their original format.

Additionally, if the configuration passed to the agent accommodates the `grep` or `grep-invert` options, corresponding attributes should be appended to the entire launch entity in ReportPortal.

### CODE REVIEW

File: src/reporter.ts [hunk 1]
```diff
@@ -94,6 +94,8 @@ export class RPReporter implements Reporter {
       extendTestDescriptionWithLastError: true,
       ...config,
       launchId: process.env.RP_LAUNCH_ID || config.launchId,
+      // TODO: consider the "grep" and "grep-invert" options
+
```
Comment (AmsterGet) [inline src/reporter.ts:98]: You can use [`onBegin`](https://playwright.dev/docs/test-reporters#custom-reporters:~:text=implements%20Reporter%20%7B-,onBegin,-(config%3A) method to access `grep` and `grepInvert` options. If they are present in `config` parameter, they should be added to the launch attributes in the next format:
- For config option `{ grep: '@fast' }` the launch attribute should be `{ key: 'grep', value: '@fast' }`;
- For config option `{ grepInvert: '@fast' }` the launch attribute should be `{ key: 'grepInvert', value: '@fast' }`.

**Review by AmsterGet — CHANGES_REQUESTED** [review]
Please also don't forget to cover your changes with tests and update readme file with new section `Attributes` under `Reporting` chapter with the guide how to use the feature.

File: src/models/configs.ts [hunk 2]
```diff
@@ -56,4 +56,6 @@ export interface ReportPortalConfig extends ClientConfig, AttachmentsConfig {
   includeTestSteps?: boolean;
   includePlaywrightProjectNameToCodeReference?: boolean;
   extendTestDescriptionWithLastError?: boolean;
+  grep?: string;
+  grepInvert?: string;
```
Comment (AmsterGet) [inline src/models/configs.ts:60]: These parameters are not the part of the agent config, they will come from the playwright configuration, so please remove it here.

File: src/reporter.ts [hunk 3]
```diff
@@ -282,6 +282,15 @@ export class RPReporter implements Reporter {
...
       id: launchId,
     };
+
+    // Extract grep and grepInvert from config and add as launch attributes
+    if (this.config.grep) {
+      startLaunchObj.attributes.push({ key: 'grep', value: this.config.grep });
+    }
+    if (this.config.grepInvert) {
+      startLaunchObj.attributes.push({ key: 'grepInvert', value: this.config.grepInvert });
+    }
```
Comment (AmsterGet) [inline src/reporter.ts:292]: Use [config: FullConfig argument of onBegin method](https://playwright.dev/docs/test-reporters#custom-reporters:~:text=implements%20Reporter%20%7B-,onBegin) to check if the `grep` or `grepInvert` options provided.

File: tests/example.spec.js
Comment (AmsterGet) [inline tests/example.spec.js:1]: We use jest to write unit tests for this agent, so please update unit tests from `__tests__`.
I suppose it should be `src/__tests__/reporter/startSuiteTestReporting.spec.ts` for `onTestBegin` and `src/__tests__/reporter/launchesReporting.spec.ts` for `onBegin`.
I suppose this file should be removed.

File: README.md [hunk 4]
```diff
@@ -1,49 +1,55 @@
...
+      key: 'key',
+      value: 'value',
+    },
+    {
+      value: 'value',
+    },
+  ],
+  description: 'Your launch description',
+  grep: 'fast', // to run only tests with 'fast' tag in the title
+  grepInvert: 'fast', // to exclude tests with 'fast' tag in the title
```
Comment (AmsterGet) [inline README.md:39]: Should be removed as it is a part of playwright configuration.

**Review by AmsterGet — CHANGES_REQUESTED** [review]

File: tests/example.spec.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+// @ts-check
```
Comment (AmsterGet) [inline tests/example.spec.js:1]: Please create a new branch in [examples repo](https://github.com/reportportal/examples-js/blob/main/example-playwright/tests/rp-features/attributes-and-description.spec.ts) as mentioned in the task and store this example there.
Also please add a `@tag` to at least one of them to demonstrate the feature.

File: src/reporter.ts [hunk 6]
```diff
@@ -635,4 +653,17 @@ export class RPReporter implements Reporter {
...
+    return (
+      title
+        .match(/@\w+\s*/g)
+        ?.join('')
+        .trim() || ''
+    );
+  }
+
+  #getAttributesFromTitle(title: string): Attribute[] {
+    return title.match(/@\w+\s*/g)?.map((tag) => ({ key: 'tag', value: tag.trim() })) || [];
```
Comment (AmsterGet) [inline src/reporter.ts:667]: As we discussed earlier, the attributes can be `@key:value` or just `@value`, so please consider the two cases here.
In case of only `value` presented, `key` should be empty.

File: src/__tests__/reporter/startSuiteTestReporting.spec.ts [hunk 7]
```diff
@@ -158,4 +159,32 @@ describe('start reporting suite/test', () => {
     expect(reporter.suites).toEqual(new Map());
     expect(reporter.testItems).toEqual(new Map());
   });
+
+  test('@smoke client.startTestItem should be called with one tag at the beginning', () => {
```
Comment (AmsterGet) [inline src/__tests__/reporter/startSuiteTestReporting.spec.ts:163]: Did you run the tests locally? I see that they are failing in CI.
Also the way its written won't work as you need to add a `@smoke` to the `simpleTestCase`'s `title` to test the feature.

**Review by AmsterGet — CHANGES_REQUESTED** [review]

File: tests/example.spec.js (hunk 5 above)
Comment (abdelrhman-arnos) [reply tests/example.spec.js:1]: It wasn't mentioned in the task to update another repository, I'll create a new pull request for these changes anyway.

Comment (AmsterGet) [reply tests/example.spec.js:1]: This may be some kind of fluke in migrating the task to the eWork platform, as we mentioned updating the examples in the acceptance criteria (see screenshot below).

image

Comment (abdelrhman-arnos) [reply tests/example.spec.js:1]: Hello @AmsterGet , please check out the new PR https://github.com/reportportal/examples-js/pull/50

File: src/__tests__/reporter/startSuiteTestReporting.spec.ts (hunk 7 above)
Comment (abdelrhman-arnos) [reply src/__tests__/reporter/startSuiteTestReporting.spec.ts]: Hi @AmsterGet ,
At present, I'm encountering an issue with the tests locally as they are all failing, including the older ones. As a result, I'm currently validating the tests on GitHub actions.

File: README.md [hunk 8]
```diff
@@ -143,11 +151,13 @@ All ReportingApi methods have an optional _suite_ parameter.<br/>
 If you want to add a data to the suite, you must pass the suite name as the last parameter.
 
 ##### addAttributes
```
Comment (AmsterGet) [inline README.md:153]: Please update this section with a small code example and mention that users can now specify test attributes in test titles.

File: src/__tests__/reporter/retriesReporting.spec.ts [hunk 9]
```diff
@@ -41,7 +41,7 @@ describe('retries reporting', () => {
   test('client.startTestItem should be called with retry=true params', () => {
     const parentId = 'tempTestItemId';
     const expectedTestObj: StartTestObjType = {
-      startTime: reporter.client.helpers.now(),
+      startTime: expect.any(Number),
```
Comment (AmsterGet) [inline src/__tests__/reporter/retriesReporting.spec.ts:44]: What is the purpose of this change?

Comment (abdelrhman-arnos) [reply src/__tests__/reporter/retriesReporting.spec.ts]: It was part of my experiments, I'll revert it.

File: README.md [hunk 10]
```diff
@@ -163,13 +171,27 @@ test('should have the correct attributes', () => {
 });
 ```
 
+You can now add test attributes in test titles using `@tag` notation. This provides a succinct way to include metadata in test reports. Any `@tag` in a jest test title will be used as an attribute in ReportPortal, but won't appear in the final test title on ReportPortal.
```
Comment (AmsterGet) [inline README.md:174]:
```suggestion
You can now add test attributes in test titles using `@tag` notation. This provides a succinct way to include metadata in test reports. Any `@tag` in a playwright test title will be used as an attribute in ReportPortal, but won't appear in the final test title on ReportPortal.
```

File: README.md [hunk 11]
```diff
@@ -163,13 +171,27 @@ test('should have the correct attributes', () => {
...
+
+Example:
+
+```javascript
+test('should have a tag at the beginning of the test title', () => {
```
Comment (AmsterGet) [inline README.md:179]: I suppose that we need to provide an example of the feature mentioned above.
```suggestion
test('@tag should have a tag at the beginning of the test title', () => {
```

File: README.md [hunk 12]
```diff
@@ -163,13 +171,27 @@ test('should have the correct attributes', () => {
...
+  ReportingApi.addAttributes([{ key: 'title', value: 'tag' }]);
```
Comment (AmsterGet) [inline README.md:180]: So please remove this method usage just to avoid mixing two approaches in this example.
```suggestion
```

File: README.md [hunk 13]
```diff
@@ -52,27 +56,30 @@ The full list of available options presented below.
...
+| grep                                        | Optional   | Not set   | To run all the test cases contains a specific tag i.e. "@fast"                                                                                                                                                                                                                                                                                                                        |
+| grepInvert                                  | Optional   | Not set   | To run all the test cases not contains a specific tag i.e. "@fast"                                                                                                                                                                                                                                                                                                                    |
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply README.md]: @abdelrhman-arnos as I see you removed the mention about `grep`s at all. Please add a short description about it to the `## Reporting` section.

[the remark this answers is not in the document]
Comment (abdelrhman-arnos) [reply README.md]: I've created a new section for the grep options associated with an example of how can the user configure it. Please re-review the changes.

**Review by AmsterGet — APPROVED** [review]

**Comment by pashidlos** [comment]
@abdelrhman-arnos thanks for the feature!
@AmsterGet when this gonna be released?
